File: ClassEval_T/cpp/solution/BoyerMooreSearch.cpp
#include <string>
#include <vector>

class BoyerMooreSearch {
public:
    BoyerMooreSearch(const std::string& text, const std::string& pattern);

    int match_in_pattern(char c) const;
    int mismatch_in_text(int currentPos) const;
    std::vector<int> bad_character_heuristic() const;

private:
    std::string text;
    std::string pattern;
    int text_len;
    int pat_len;
};

BoyerMooreSearch::BoyerMooreSearch(const std::string& text, const std::string& pattern)
    : text(text), pattern(pattern), text_len(text.size()), pat_len(pattern.size()) {}

int BoyerMooreSearch::match_in_pattern(char c) const {
    for (int i = pat_len - 1; i >= 0; --i) {
        if (c == pattern[i]) {
            return i;
        }
    }
    return -1;
}

int BoyerMooreSearch::mismatch_in_text(int currentPos) const {
    for (int i = pat_len - 1; i >= 0; --i) {
        if (pattern[i] != text[currentPos + i]) {
            return currentPos + i;
        }
    }
    return -1;
}

std::vector<int> BoyerMooreSearch::bad_character_heuristic() const {
    std::vector<int> positions;
    int i = 0;
    while (i <= text_len - pat_len) {
        int mismatch_index = mismatch_in_text(i);
        if (mismatch_index == -1) {
            positions.push_back(i);
            i += 1;  
        }
        else {
            char mismatched_char = text[mismatch_index];
            int match_index = match_in_pattern(mismatched_char);
            if (match_index == -1) {
                i += (mismatch_index - i + 1);
            }
            else {
                i += (mismatch_index - match_index);
            }
        }
    }
    return positions;
}

File: ClassEval_T/java/solutuon/BoyerMooreSearch.java
package org.example;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class BoyerMooreSearch {
    private final String text;
    private final String pattern;
    private final int textLen;
    private final int patLen;

    public BoyerMooreSearch(String text, String pattern) {
        this.text = text;
        this.pattern = pattern;
        this.textLen = text.length();
        this.patLen = pattern.length();
    }

    int matchInPattern(char ch) {
        return pattern.lastIndexOf(ch);
    }

    int mismatchInText(int currentPos) {
        for (int i = patLen - 1; i >= 0; i--) {
            if (pattern.charAt(i) != text.charAt(currentPos + i)) {
                return currentPos + i;
            }
        }
        return -1;
    }

    public List<Integer> badCharacterHeuristic() {
        List<Integer> positions = new ArrayList<>();
        int i = 0;

        Map<Character, Integer> badCharHeuristic = new HashMap<>();
        for (int j = 0; j < patLen; j++) {
            badCharHeuristic.put(pattern.charAt(j), j);
        }

        if(patLen==0)
        {
            for(int j=0;j<=textLen;++j)
            {
                positions.add(j);
            }
            return positions;
        }

        while (i <= textLen - patLen) {
            int mismatchIndex = mismatchInText(i);
            if (mismatchIndex == -1) {
                positions.add(i);
                i += patLen;
            } else {
                char mismatchChar = text.charAt(mismatchIndex);
                int matchIndex = badCharHeuristic.getOrDefault(mismatchChar, -1);
                if (matchIndex >= 0) {
                    i += Math.max(1, mismatchIndex - i - matchIndex);
                }
                else {
                    i += mismatchIndex - i + 1;
                }
            }
        }
        return positions;
    }
}
